worldedit-core/src/main/java/com/sk89q/worldedit/internal/util/RecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,234 @@
+                            continue;
+                        }
+                        // make sure to work with an absolute path
+                        Path path = (Path) event.context();
+                        Path parentPath = watchRootMap.get(watchKey);
+                        path = parentPath.resolve(path);
+
+                        if (kind.equals(StandardWatchEventKinds.ENTRY_CREATE)) {
+                            path = WorldEdit.getInstance().getSafeOpenFile(null, schematicRoot.toFile(), schematicRoot.relativize(path).toString(), null).toPath();
+
+                            if (Files.isDirectory(path)) { // new subfolder created, create watch for it
+                                try {
+                                    registerFolderWatcher(path);
+                                    triggerInitialEvents(path);
+                                } catch (IOException | FilenameException e) {
+                                    LOGGER.error(e);
+                                }
+                            } else { // new file created
+                                eventConsumer.accept(new FileCreatedEvent(path));
+                            }
+                        } else if (kind.equals(StandardWatchEventKinds.ENTRY_DELETE)) {
+                            // When we are notified about a deleted entry, we can't simply ask the filesystem
+                            // whether the entry is a file or a folder. But we have our watchRootMap, that stores
+                            // one WatchKey per (sub)folder, so we can just ask it.
+                            if (watchRootMap.containsValue(path)) { // was a folder
+                                LOGGER.debug("Watch unregistered: " + path);
+                                WatchKey obsoleteSubfolderWatchKey = watchRootMap.inverse().get(path);
+                                // stop listening to changes from deleted dir
+                                obsoleteSubfolderWatchKey.cancel();
+                                watchRootMap.remove(obsoleteSubfolderWatchKey);
+                                eventConsumer.accept(new DirectoryDeletedEvent(path));
+                            } else { // was a file
+                                eventConsumer.accept(new FileDeletedEvent(path));
+                            }
+                        }
+                    }
+
+                    if (!watchKey.reset()) {
+                        watchRootMap.remove(watchKey);
+                        if (watchRootMap.isEmpty()) {
+                            break; // nothing left to watch
+                        }
+                    }
+                }
+            } catch (ClosedWatchServiceException | FilenameException ignored) {
+            }
+            LOGGER.debug("RecursiveDirectoryWatcher::EventConsumer exited");
+        });
+        watchThread.setName("RecursiveDirectoryWatcher");
```

Comment:
At the very least we should identify ourselves fully. Perhaps we should also use a global counter.
```suggestion
        watchThread.setName("WorldEdit-RecursiveDirectoryWatcher");
```